### PR TITLE
#214 add minimal detail page components

### LINKED ISSUE #214: Add minimal entity details interface
Add minimal entity details UI.

## Expected Behaviour

See invision design.

## Current Behaviour

N/A.

## Possible Solution

- Display entity code.
- Display entity description.
- Display entity children.
- Display entity properties.
- Display entity tree, e.g. `Amoxicillin 250mg / Oral / Capsule`.

### PR DESCRIPTION
**Pull request by wlthomson**
Fixes #214.

Branches off #222.

## Description

Adds minimal details page.

### Documentation:

Tick below if one of the following applies:

1. Documentation has been added to cover changes introduced by this PR.
2. This PR requires no changes to documentation.

- [ ] Documentation OK.

### Tests:

Tick below if one of the following applies:

1. Tests have been added to cover changes introduced by this PR.
2. This PR requires no new tests.

- [ ] Tests OK.

### CODE REVIEW

File: unified-codes/apps/web/src/components/layout/ErrorLayout.tsx
Comment (wlthomson) [inline unified-codes/apps/web/src/components/layout/ErrorLayout.tsx:1]: A bit unnecessary here, but for the sake of consistency and to keep a common pattern going! Will probably find a better layout when more error pages are added.

File: unified-codes/libs/ui/src/lib/hooks/useToggle.tsx
Comment (wlthomson) [inline unified-codes/libs/ui/src/lib/hooks/useToggle.tsx:21]: Completely unneeded... but cool!

File: unified-codes/apps/web/src/components/containers/error/ErrorPage.tsx
Comment (wlthomson) [inline unified-codes/apps/web/src/components/containers/error/ErrorPage.tsx:58]: Would be good to add some more error pages, e.g. for authentication errors. Something for a separate issue?

**Review by wlthomson — COMMENTED** [review]
Went a bit off-road with a few fix ups. I think this finishes off all the refactoring so this will be the last of the big ones!

File: unified-codes/apps/web/src/components/containers/error/ErrorPage.tsx
Comment (mark-prins) [reply unified-codes/apps/web/src/components/containers/error/ErrorPage.tsx:58]: Yes, absolutely. I'd be keen to throw a `418` error - and perhaps there's already a good gif for that one?
Server error could be trickier, as `nginx` should raise / handle that. but a 403, surely? perhaps when we do admin - just to keep this PR moving along.

**Comment by mark-prins** [comment]
When running this, I had to remove `/entity` from https://github.com/openmsupply/unified-codes/blob/%23214-add-minimal-detail-interface/unified-codes/libs/ui/src/lib/components/index.ts

File: unified-codes/apps/web/src/components/containers/detail/DetailEntityListItem.tsx
Comment (katherine-sussol) [inline unified-codes/apps/web/src/components/containers/detail/DetailEntityListItem.tsx:48]: 😻 🌳

File: unified-codes/apps/web/src/components/containers/detail/DetailEntityList.tsx
Comment (mark-prins) [inline unified-codes/apps/web/src/components/containers/detail/DetailEntityList.tsx:29]: Would be nice to have a 'no results' option here too - if there are no children to display

File: unified-codes/apps/web/src/components/containers/error/ErrorPage.tsx
Comment (mark-prins) [inline unified-codes/apps/web/src/components/containers/error/ErrorPage.tsx:34]: it makes me think of a `sigh`.. should really be changed though. The text alignment on the bottom is not centered

File: unified-codes/apps/web/src/components/containers/app/AppPage.tsx
Comment (mark-prins) [inline unified-codes/apps/web/src/components/containers/app/AppPage.tsx:63]: This one should be redirecting to `/error/404` - as it stands the user gets routed to the URL `http://localhost:4200/error/:code`

File: unified-codes/apps/web/src/components/containers/error/ErrorPage.tsx
Comment (mark-prins) [inline unified-codes/apps/web/src/components/containers/error/ErrorPage.tsx:58]: I think the `ErrorNotFoundTop` `ErrorNotFoundCenter` and `ErrorNotFoundBottom` are OTT - why not just have an `ErrorNotFoundPage`? Or even an error page where the image and messaging is specific to the error and sourced somehow?

**Review by katherine-sussol — COMMENTED** [review]
Think I faced the same thing @mark-prins mentioned, commented out entity to get the PR running.

Otherwise, I think it's looking pretty neat, a couple of minor UX things (but don't think these should necessarily block the PR! Could do them in separate issues):
- Maybe when there are no properties the property list shouldn't be clickable? Or not rendered? Or...(insert alternative suggestion/rebuttal here)?
- Sometimes when there are a lot of items and everything is fully expanded items are pushed outside the viewable area. E.g. 
Screen Shot 2020-10-14 at 4 15 40 PM

File: unified-codes/apps/web/src/components/containers/error/ErrorNotFoundCenter.tsx
Comment (mark-prins) [inline unified-codes/apps/web/src/components/containers/error/ErrorNotFoundCenter.tsx:26]: wait, what? `Centre` surely?

File: unified-codes/apps/web/src/components/containers/explorer/ExplorerTableRows.tsx
Comment (mark-prins) [inline unified-codes/apps/web/src/components/containers/explorer/ExplorerTableRows.tsx:52]: nice!

File: unified-codes/apps/web/src/components/layout/ErrorLayout.tsx
Comment (mark-prins) [reply unified-codes/apps/web/src/components/layout/ErrorLayout.tsx:1]: Oh soz. I whined about this elsewhere.
Yes, agree re the consistency - though I would have ditched consistentness in this case, it works and is elegant. Ignore previous comment 😉

**Review by mark-prins — CHANGES_REQUESTED** [review]
There's a couple of minor things that I'd like to see changed - as ever though, the code is beautiful and thoughtfully crafted.

**Comment by mark-prins** [comment]
thanks @katherine-sussol - I was going to mention the same things that you have 😄 

Some possible tidy-up-inna-new-issue items:
- the circular dependency warnings
- layout tweaks to the details page
- more errors!
- pagination no longer works

The first and last of these I fixed in the deployment branch, so I could merge those changes on through after this PR is merged.

Re the details page, once piece of feedback that I've received is that it wasn't obvious that you could click on 'Form' or 'Properties'.. I thought it was obvious enough myself. I did find the multiple clicking annoying - there's not a lot of information, so perhaps some of it could be expanded / all laid out. Though I don't want to remove the recursive list clicking thing! that has to stay, somehow.

File: unified-codes/apps/web/src/components/containers/error/ErrorPage.tsx
Comment (wlthomson) [reply unified-codes/apps/web/src/components/containers/error/ErrorPage.tsx:58]: I agree 100%... that was what I did first! I was worried about confusing people since we've implemented a bit of a page convention. I was hoping that as we add more error pages we will find that they share a common layout structure (e.g. maybe each error will have a shared image but different text or something). 

The reason I avoided `ErrorXPage` was to avoid confusion with the "page" convention (i.e. a page is something which is routed to from the main app and which has `onMount` and `onUnmount` props which take the navigation props). But I think as I look at it again, it's probably best to just has specific error pages.

File: unified-codes/apps/web/src/components/containers/error/ErrorNotFoundCenter.tsx
Comment (wlthomson) [reply unified-codes/apps/web/src/components/containers/error/ErrorNotFoundCenter.tsx:26]: I've been beaten down by the Americanization of everything browser/CSS related. Even if I revert back to `"center"`, the commit history is all that's needed for them to lock my up in the tower. 

@katherine-sussol @mark-prins Front-end spelling conventions: American English or correct English?

File: unified-codes/apps/web/src/components/containers/error/ErrorPage.tsx
Comment (wlthomson) [reply unified-codes/apps/web/src/components/containers/error/ErrorPage.tsx:58]: I think maybe for now we put up with this agreed OTT solution, and maybe when more error components are added, we can either adjust the layout to be a bit less redundant, or revert to having a `ErrorXPage` for each error page?

File: unified-codes/apps/web/src/components/containers/error/ErrorNotFoundCenter.tsx
Comment (mark-prins) [reply unified-codes/apps/web/src/components/containers/error/ErrorNotFoundCenter.tsx:26]: 🤣  that sounded like a loaded question!
I'm a pedant, and I run with `correct English` - really doesn't trouble me either way though

File: unified-codes/apps/web/src/components/containers/error/ErrorPage.tsx
Comment (mark-prins) [reply unified-codes/apps/web/src/components/containers/error/ErrorPage.tsx:58]: yup, sounds good to me

File: unified-codes/apps/web/src/components/containers/error/ErrorNotFoundCenter.tsx
Comment (katherine-sussol) [reply unified-codes/apps/web/src/components/containers/error/ErrorNotFoundCenter.tsx:26]: I also prefer correct English! We mustn't give in to these external pressures to Americanize!
... Actually I don't mind either and tailor depending on audience. But I think our audience here is !🇺🇸

File: unified-codes/apps/web/src/components/containers/app/AppPage.tsx
Comment (wlthomson) [reply unified-codes/apps/web/src/components/containers/app/AppPage.tsx]: Oops! 3b2705216e6e80f418c76bfa4ee81f2d4834e3c7.

File: unified-codes/apps/web/src/components/containers/error/ErrorPage.tsx
Comment (wlthomson) [reply unified-codes/apps/web/src/components/containers/error/ErrorPage.tsx]: 3b2705216e6e80f418c76bfa4ee81f2d4834e3c7.

File: unified-codes/apps/web/src/components/containers/app/AppPage.tsx
Comment (wlthomson) [reply unified-codes/apps/web/src/components/containers/app/AppPage.tsx:43]: To be fixed in standalone issue!

File: unified-codes/apps/web/src/components/containers/detail/DetailEntityList.tsx
Comment (wlthomson) [reply unified-codes/apps/web/src/components/containers/detail/DetailEntityList.tsx:37]: For now just not rendering, but will need to address this in a separate issue (a bit tricky to handle these sort of cases with recursion... i.e. we don't want the leafs also displaying "no results", so will have to think about this one a bit).

File: unified-codes/apps/web/src/components/containers/error/ErrorPage.tsx
Comment (wlthomson) [reply unified-codes/apps/web/src/components/containers/error/ErrorPage.tsx]: 3b2705216e6e80f418c76bfa4ee81f2d4834e3c7.

File: unified-codes/apps/web/src/components/containers/app/AppPage.tsx
Comment (wlthomson) [reply unified-codes/apps/web/src/components/containers/app/AppPage.tsx]: 3b2705216e6e80f418c76bfa4ee81f2d4834e3c7.

**Comment by wlthomson** [comment]
@mark-prins @katherine-sussol If you're both happy to merge, I'm happy to address the remaining stuff in new issues:

Bug fixes:

- Circular depedencies.
- Pagination bugs.
- Missing key errors.

Features:

- Error handling for navigating to detail page with non-existent entity code.
- Loading spinner for details page.
- Handling non-drug entities in details page.
- Handling empty child/property lists.

**Review by mark-prins — APPROVED** [review]

**Pull request merged**
